### PR TITLE
fix(errorHandler): async error handling for watchers

### LINKED ISSUE #9625: Vue.config.errorHandler cannot catch the error thrown by watcher whose user is true in the option if the callback returns a promise
### Version
2.6.8

### Reproduction link
[https://jsfiddle.net/fh5kz3r6/](https://jsfiddle.net/fh5kz3r6/)

### Steps to reproduce
1.return a promise that reject one error in watch;
2.trigger the watcher in the step 1;

### What is expected?
Vue.config.errorHandler can catch the error.

### What is actually happening?
Vue.config.errorHandler not catch the error.

---
I think changing the try {/.../} catch (e) {/.../}  in the Vue.prototype.$watch and run function of the watcher to invokeWithErrorHandling can solve the problem.

### LINKED ISSUE #10009: Vue.config.errorHandler and errorCaptured should capture rejected promised in watchers
### Version
2.6.6

### Reproduction link
Vue.config.errorHandler: [https://jsfiddle.net/0aeufxmr/5/](https://jsfiddle.net/0aeufxmr/5/)
`errorCaptured` option: https://jsfiddle.net/shenron/kbuymoz0/

### Steps to reproduce
Just click `click1` and `click2` you will see the difference

### What is expected?
when clicking `click1`, the message should be error response

### What is actually happening?
message not changed

### PR DESCRIPTION
**Pull request by hello-brsd**
Added async error handling for watchers and immediate watchers

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
Fixes #9625 
Fixes #10009

### CODE REVIEW

**Comment by jfgodoy** [comment]
I'm waiting for this.
@hello-brsd you should squash your commits into one

**Comment by posva** [comment]
no need to squash your commits, we already do when merging 😉

**Comment by hello-brsd** [comment]
@posva Thank you for the review, i added some extra test case to the `errorCaptured.spec.js`.

File: test/unit/features/options/errorCaptured.spec.js [hunk 1]
```diff
@@ -247,4 +247,148 @@ describe('Options errorCaptured', () => {
...
+
+    new Vue({
+      errorCaptured: spy,
+      render: h => h(Child),
+    }).$mount()
+
+    waitForUpdate(() => {
+      expect(spy).toHaveBeenCalledWith(err, child, 'callback for watcher "foo"')
+      expect(globalSpy).toHaveBeenCalledWith(err, child, 'callback for watcher "foo"')
+    }).then(Vue.nextTick(done))
```
Comment (posva) [inline test/unit/features/options/errorCaptured.spec.js]: Shouldn't this be something like

```suggestion
    }).then(() => Vue.nextTick(done))
```

or `then(done)`

Right now the code is equivalent to

```js
    waitForUpdate(() => {
	      expect(spy).toHaveBeenCalledWith(err, child, 'callback for watcher "foo"')
	      expect(globalSpy).toHaveBeenCalledWith(err, child, 'callback for watcher "foo"')
	    })
     Vue.nextTick(done)
```

Comment (hello-brsd) [reply test/unit/features/options/errorCaptured.spec.js]: Oops, sorry about that.

Comment (hello-brsd) [reply test/unit/features/options/errorCaptured.spec.js]: I fixed it.

**Comment by abhipanda** [comment]
@posva - Is there an ETA for this merge ?

**Comment by Ishadimu** [comment]
Great work @hello-brsd! Until this is merged, I'm having to write an interim solution using axios interceptors along with vuex. 😬 

@posva Thanks for approving the PR. Any updates on merge ETA?

**Comment by posva** [comment]
Sorry, I cannot provide any ETA so please don't ping me for it 😓

**Comment by caiquecastro** [comment]
Any updates?

**Comment by mvirkkunen** [comment]
I would also like to know what the status is on this because I've been bitten by this multiple times by now, and the fix seems very straightforward. I see Vue 2.x hasn't had a release with anything besides bug fixes for a while, so is this something you would even consider merging anymore? Or is all the focus on Vue 3 now?

**Comment by rogierpennick** [comment]
Any updates on this? Our error handling would be greatly improved by this because we use a lot of async watches in our code. I'd really like to avoid having to resort to workarounds!

**Comment by OzCroot** [comment]
Can this please be reviewed?

**Comment by msokk** [comment]
@OzCroot Looks like this is already approved a while ago, @posva any newer ETA for this to land?

[Documentation for `errorHandler` says](https://vuejs.org/v2/api/#errorHandler): 
> Assign a handler for uncaught errors during component render function and **watchers**. 

and there is a `2.6.0+` update section that mentions added async support:
> In addition, if any of the covered hooks or handlers returns a Promise chain (e.g. async functions), the error from that Promise chain will also be handled.

Overall one would expect from documentation that async watchers would work with global error handler, but currently it is not so 🙂

**Comment by berniegp** [comment]
Here`s a fix I came up with until this PR is merged.

    // Fix "Vue.config.errorHandler and errorCaptured should capture rejected promised in watchers"
    // https://github.com/vuejs/vue/issues/10009
    // Ugly monkey patching until https://github.com/vuejs/vue/pull/9484 is merged
    const realWatch = Vue.prototype.$watch;
    Vue.prototype.$watch = function $watchDetour(expOrFn, callback, options) {
      if (typeof callback === 'function') {
        if (this.$_watchCount === undefined) {
          this.$_watchCount = 0;
        }

        // Reroute $watch to $on which handles promise rejections properly
        const watchDetourEvent = `watchDetour_${this.$_watchCount}`;
        this.$_watchCount += 1;
        const realCallback = callback;
        this.$on(watchDetourEvent, (args) => { return realCallback.apply(this, args); });
        callback = (...args) => { this.$emit(watchDetourEvent, args); };
      }

      // realWatch will call back into $watchDetour with a function if callback wasn't one
      return realWatch.call(this, expOrFn, callback, options);
    };

**Pull request merged**